packages/middleware/src/loggerManager.ts

```diff
@@ -0,0 +1,108 @@
+import type { Response } from "express";
+import { LoggerInterface } from "@vue-storefront/logger";
+import type {
+  AlokaiContainer,
+  IntegrationContext,
+  MiddlewareConfig,
+} from "./types";
+
+type LoggerInstance = LoggerInterface;
+const GLOBAL_MIDDLEWARE_CFG_KEY = Symbol("GLOBAL_MIDDLEWARE_CFG_KEY");
+export class LoggerManager<
+  TLoggerConfig = any,
+  TIntegrationContext extends Record<string, IntegrationContext> = any
+> {
+  private instances: Record<string | symbol, LoggerInstance> = {};
+
+  /** Creating minimal amount of needed Logger instances * */
+  constructor(
+    config: MiddlewareConfig<TIntegrationContext>, // Whole middleware.config.js object
+    buildLogger: (logger: TLoggerConfig) => LoggerInstance
+  ) {
+    // Global Logger
+    this.instances[GLOBAL_MIDDLEWARE_CFG_KEY] = buildLogger(
+      config.logger as TLoggerConfig // logger is on same level as integrations
+    );
+    // Logger for every integration having custom logger config
+    for (const [integrationName, integrationEntry] of Object.entries(
+      config.integrations
+    )) {
+      if (this.hasCustomLoggerConfig(integrationEntry)) {
+        // every integration inside `configuration` object can have own logger too
+        this.instances[integrationName] = buildLogger(
+          integrationEntry.logger as TLoggerConfig
+        );
+      }
+    }
+  }
+
+  // eslint-disable-next-line class-methods-use-this
+  private hasCustomLoggerConfig(integrationEntry) {
+    return Boolean(integrationEntry.logger);
+  }
+
+  // getting reference to instance of logger by integration tag name
+  public get(integrationName) {
+    return (
+      this.instances[integrationName] ||
+      this.instances[GLOBAL_MIDDLEWARE_CFG_KEY]
+    );
+  }
+
+  // getting reference to global instance of logger
+  public getGlobal() {
+    return this.instances[GLOBAL_MIDDLEWARE_CFG_KEY];
+  }
+}
+
+export function getLogger(
+  source: AlokaiContainer | Response | { res: Response }
+): LoggerInstance {
+  if (!source) {
+    throw new Error("Parameter for getLogger function not provided");
+  }
+  if ("logger" in source) {
+    return source.logger;
+  }
+  let base: Response;
+  if ("res" in source) {
+    base = source.res;
+  } else {
+    base = source;
+  }
+  return base.locals.alokai.logger;
+}
+
+const methodsToOverwrite = [
+  "emergency",
+  "alert",
+  "critical",
+  "error",
+  "warning",
+  "notice",
+  "info",
+  "debug",
+];
+
+export function wrapLogger(
+  logger: LoggerInstance,
+  metadataGetter: (metadata: Record<string, any>) => Record<string, any>
+): LoggerInstance {
+  return new Proxy(logger, {
+    get(target, prop) {
+      if (
+        typeof target[prop] === "function" &&
+        methodsToOverwrite.includes(prop as string)
```

Comment:
We should loop over all methods not just hardcoded set of names, we can't miss something if someone will add new one. This will not work for him.

Reply:
The overwritten method has to match the interface anyway. I've just found an exposed `log` method that doesn't match it. So I reverted logic to specify methods to skip.

```ts
import type { LoggerInterface } from "@vue-storefront/logger";

const METHODS_TO_SKIP = ["log"];

export function injectMetadata(
  logger: LoggerInterface,
  metadataGetter: (metadata: Record<string, any>) => Record<string, any>
): LoggerInterface {
  return new Proxy(logger, {
    get(target, prop) {
      const shouldSkipMethod =
        typeof target[prop] !== "function" ||
        METHODS_TO_SKIP.includes(prop as string);

      if (!shouldSkipMethod) {
        return (...args: any[]) => {
          const [message, metadata] = args;
          target[prop](message, {
            ...metadata,
            ...metadataGetter(metadata),
          });
        };
      }
      return target[prop];
    },
  });
}
```

Reply:
Yes, this is also what I have proposed in the SDK, to have blacklist rather than whitelist. At some point we might want to expose it as a configurable options then it should be a shorter list than whitelist.